NIFIREG-313 - Add OpenId Connect support for authenticating users

thenatog opened the pull request:
NIFIREG-313 - Added Login and Logout filters. Added unit tests.

NIFIREG-313 - Added changes from NiFi PR #4344. Fixing tests.

NIFIREG-313 - alopresto provided a patch to fix failing tests as a result of the NiFiRegistryProperties mocking not working the same as NiFi and other minor issues.

NIFIREG-313 - Added the OIDC identity provider to the security config.

NIFIREG-313 - Testing javascript to make call to exchange OIDC cookie for JWT. I believe some parts in the NiFiRegistrySecurityConfig are not required, similar to how Kerberos auth works.

NIFIREG-313 - OIDC Authentication works correctly. Need to adjust javascript to allow either Kerberos or OIDC - currently only allows OIDC.

NIFIREG-313 - Updated ticketExchange function to call both Kerberos and OIDC endpoints to retrieve a JWT.

NIFIREG-313 - User can hit the LOGIN button to log in with OIDC, or it will redirect to the default login dialog for basic credential login.

NIFIREG-313 - Removing unused imports

NIFIREG-313 - OIDC login can only be attempted if OIDC is configured.

NIFIREG-313 - Fixed ticketExchange() javascript to work with x509 certs. Removed unnecessary code from LogoutFilter.

Thank you for submitting a contribution to Apache NiFi Registry.

Please provide a short description of the PR here:

#### Description of PR

_Enables X functionality; fixes bug NIFIREG-YYYY._

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFIREG-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi-registry` folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] Have you verified that the full build is successful on JDK 8?
- [ ] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-registry-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-registry-assembly`?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

thenatog commented:
This change matches, as close as possible, the OIDC authentication code in NiFi. When opening the NiFi Registry UI with OIDC configured, there should be a Login button in the top right corner. Once clicked, NiFi Registry should redirect you to the configured OIDC provider login page, you can enter your credentials, and the provider should redirect back to NiFi Registry and show your logged in user in the top right. Hitting log out will log out your user from NiFi Registry and remove the user's JWT.

I have tested:
- Using X509 without any other authentication mechanism enabled.
- Using OIDC authentication provider (Google Suite). Logging in/logging out is working.
- Using LDAP through the NiFi Registry basic credentials dialog.

I think the reviewer should test using Kerberos to ensure everything is still fine there - I do not have a good set up to test that.

bbende commented on nifi-registry-core/nifi-registry-web-api/src/main/java/org/apache/nifi/registry/web/security/NiFiRegistrySecurityConfig.java:
```diff
@@ -88,7 +88,8 @@ public NiFiRegistrySecurityConfig() {
         // allow any client to access the endpoint for logging in to generate an access token
-        webSecurity.ignoring().antMatchers( "/access/token/**");
+        webSecurity.ignoring().antMatchers("/access/config", "/access/token", "/access/kerberos",
+                "/access/oidc/exchange", "/access/oidc/callback", "/access/oidc/request", "/access/token/identity-provider");
```
I'm not sure we have "/access/config", I only see "/config" which maps the ConfigResource and returns information about which type of user-group-provider and authorizer are available, which could probably go either way on whether that is exposed to unauthenticated users. I don't think it is needed by anything until after you login, but its also not exposing anything important.

thenatog replied:
Yeah good point, I'll verify all of these patterns as some appear different to NiFi. We want this list to be concise.

thenatog commented on nifi-registry-core/pom.xml:
```diff
@@ -143,6 +143,17 @@
+                <groupId>com.google.guava</groupId>
+                <artifactId>guava</artifactId>
+                <version>28.0-jre</version>
+            </dependency>
```
I'd say yes. Are there any issues with adding this dependency to the core pom? I believe this is used already in NiFi. 
Guava provides a cache used by the OIDC service.

bbende commented on nifi-registry-core/nifi-registry-properties/src/main/java/org/apache/nifi/registry/properties/NiFiRegistryProperties.java:
```diff
@@ -87,6 +92,19 @@
+    public static final String SECURITY_USER_OIDC_CLIENT_SECRET = "nifi.registry.security.user.oidc.client.secret";
+    public static final String SECURITY_USER_OIDC_PREFERRED_JWSALGORITHM = "nifi.registry.security.user.oidc.preferred.jwsalgorithm";
+    public static final String SECURITY_USER_OIDC_ADDITIONAL_SCOPES = "nifi.registry.security.user.oidc.additional.scopes";
+    public static final String SECURITY_USER_OIDC_CLAIM_IDENTIFYING_USER = "nifi.registry.security.user.oidc.claim.identifying.user";
```
The default nifi-registry.properties file needs to be updated to include these

bbende commented:
Updates look good, still two remaining comments...

1) The default nifi-registry.properties should be updated to include the OIDC properties

https://github.com/apache/nifi-registry/blob/main/nifi-registry-core/nifi-registry-resources/src/main/resources/conf/nifi-registry.properties

2) The main LICENSE and/or NOTICE files in nifi-registry-assembly should be updated to account for any new dependencies. It looks like NiFi's NOTICE file lists Guava and Nimbus, so those should be added to the NOTICE here since most likely they aren't there yet.

thenatog commented:
Added the OIDC properties and updated the NOTICE file. I'm noticing issues with the checks failing - I'm not seeing this happening locally. It also appears to be happening on Kevin's PR so perhaps these failures are unrelated to my PR.

pvillard31 commented:
Fixing the build issue in #303.

bbende commented:
+1 latest updates look good, going to merge
